crates/bevy_ecs/src/relationship/mod.rs

```diff
@@ -0,0 +1,262 @@
+/// "source" entities that relate to the given "target"
+///
+/// The [`Relationship`] component is the "source of truth" and the [`RelationshipSources`] component reflects that source of truth. When a [`Relationship`]
+/// component is inserted on an [`Entity`], the corresponding [`RelationshipSources`] component is immediately inserted on the target component if it does
+/// not already exist, and the "source" entity is automatically added to the [`RelationshipSources`] collection (this is done via "component hooks").
+///
+/// A common example of a [`Relationship`] is the parent / child relationship. Bevy ECS includes a canonical form of this via the [`Parent`](crate::hierarchy::Parent)
+/// [`Relationship`] and the [`Children`](crate::hierarchy::Children) [`RelationshipSources`].
+///
+/// [`Relationship`] and [`RelationshipSources`] should always be derived to ensure the hooks are set up properly. They will both automatically
+/// implement [`Component`] with the necessary configuration to drive the [`Relationship`].
+///
+/// ```
+/// # use bevy_ecs::relationship::{Relationship, RelationshipSources};
+/// # use bevy_ecs::entity::Entity;
+/// #[derive(Relationship)]
+/// #[relationship(relationship_sources = Children)]
+/// pub struct Parent(pub Entity);
+///
+/// #[derive(RelationshipSources)]
+/// #[relationship_sources(relationship = Parent)]
+/// pub struct Children(Vec<Entity>);
+/// ```
+///
+/// When deriving [`RelationshipSources`] you can specify the `#[relationship_sources(despawn_descendants)]` attribute to
+/// automatically despawn entities stored in an entity's [`RelationshipSources`] when that entity is despawned:
+///
+/// ```
+/// # use bevy_ecs::relationship::{Relationship, RelationshipSources};
+/// # use bevy_ecs::entity::Entity;
+/// #[derive(Relationship)]
+/// #[relationship(relationship_sources = Children)]
+/// pub struct Parent(pub Entity);
+///
+/// #[derive(RelationshipSources)]
+/// #[relationship_sources(relationship = Parent, despawn_descendants)]
+/// pub struct Children(Vec<Entity>);
+/// ```
+///
+// NOTE: The "deprecated" state is used to prevent users from mutating the internal RelationshipSource collection.
+// These internals are allowed to modify the internal RelationshipSource collection.
+#[allow(deprecated)]
+pub trait Relationship: Component + Sized {
+    /// The [`Component`] added to the "target" entities of this [`Relationship`], which contains the list of all "source"
+    /// entities that relate to the "target".
+    type RelationshipSources: RelationshipSources<Relationship = Self>;
+
+    /// Gets the [`Entity`] ID of the related entity.
+    fn get(&self) -> Entity;
+
+    /// Creates this [`Relationship`] from the given `entity`.
+    fn from(entity: Entity) -> Self;
+
+    /// The `on_insert` component hook that maintains the [`Relationship`] / [`RelationshipSources`] connection.
+    fn on_insert(mut world: DeferredWorld, entity: Entity, _: ComponentId) {
```

Comment:
This feels like a weird place for this method. I'd kinda prefer making `on_insert` etc methods for `Component` (and removing the awkward `register_component_hooks` API), and then generating those methods in the `Component` trait with the `Relationship` derive macro.

Reply:
Tracked in #17411.